File: plugins/capacitor-gr-mjpeg/ios/Sources/GrMjpegPlugin/MjpegParser.swift
import Foundation

/// Incremental MJPEG frame splitter (same algorithm as the web app's liveStream.ts and the
/// Android MjpegParser). Prefers the multipart part's Content-Length when its headers end
/// exactly where the JPEG starts; otherwise cuts on SOI (FF D8) … EOI (FF D9).
/// Not thread-safe — use from one queue.
final class MjpegParser {
    static let maxFrame = 4 * 1024 * 1024
    private static let headerWindow = 256

    private var buf: [UInt8] = []
    private var start = -1
    private var scanFrom = 0

    init() { buf.reserveCapacity(256 * 1024) }

    func reset() {
        buf.removeAll(keepingCapacity: true)
        start = -1
        scanFrom = 0
    }

    func feed(_ data: Data, onFrame: (Data) -> Void) {
        buf.append(contentsOf: data)
        var consumed = 0

        while true {
            if start < 0 {
                start = find(0xD8, from: max(consumed, scanFrom))
                if start < 0 {
                    // Keep a tail: FF D8 may straddle chunks, and the part headers that
                    // precede the next JPEG must survive until it arrives.
                    consumed = max(consumed, buf.count - Self.headerWindow)
                    scanFrom = max(consumed, buf.count - 1)
                    break
                }
            }

            let declared = partLength(start)
            if declared > 0 {
                if buf.count < start + declared { break }
                onFrame(Data(buf[start..<(start + declared)]))
                consumed = start + declared
                start = -1
                scanFrom = consumed
                continue
            }

            let end = find(0xD9, from: max(start + 2, scanFrom))
            if end < 0 {
                scanFrom = max(start + 2, buf.count - 1)
                if buf.count - start > Self.maxFrame { // garbage — resync
                    consumed = buf.count
                    start = -1
                    scanFrom = buf.count
                }
                break
            }
            onFrame(Data(buf[start..<(end + 2)]))
            consumed = end + 2
            start = -1
            scanFrom = consumed
        }
        compact(consumed)
    }

    private func compact(_ consumed: Int) {
        guard consumed > 0 else { return }
        buf.removeFirst(min(consumed, buf.count))
        if start >= 0 { start -= consumed }
        scanFrom = max(0, scanFrom - consumed)
    }

    private func find(_ second: UInt8, from: Int) -> Int {
        let n = buf.count
        var i = max(0, from)
        return buf.withUnsafeBufferPointer { p -> Int in
            while i < n - 1 {
                if p[i] == 0xFF && p[i + 1] == second { return i }
                i += 1
            }
            return -1
        }
    }

    /// Content-Length from part headers ending ("\r\n\r\n") exactly where the JPEG starts.
    private func partLength(_ jpegStart: Int) -> Int {
        guard jpegStart >= 16 else { return 0 }
        // Compare bytes: in a Swift String "\r\n" is ONE Character, so character math is wrong here.
        guard buf[jpegStart - 4] == 13, buf[jpegStart - 3] == 10,
              buf[jpegStart - 2] == 13, buf[jpegStart - 1] == 10 else { return 0 }
        let from = max(0, jpegStart - Self.headerWindow)
        let head = String(decoding: buf[from..<jpegStart], as: Unicode.ASCII.self)
        let partStart = head.range(of: "--", options: .backwards)?.lowerBound ?? head.startIndex
        let part = head[partStart...].lowercased()
        guard let key = part.range(of: "content-length:") else { return 0 }
        let digits = part[key.upperBound...].drop { $0 == " " }.prefix { $0.isNumber }
        guard let value = Int(digits), value > 0, value < Self.maxFrame else { return 0 }
        return value
    }
}

File: plugins/capacitor-gr-mjpeg/android/src/main/java/com/grremote/plugins/grmjpeg/MjpegParser.java
package com.grremote.plugins.grmjpeg;

import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.Locale;

/**
 * Incremental MJPEG frame splitter (same algorithm as the web app's liveStream.ts).
 *
 * Feed it network chunks; it emits complete JPEGs. Uses the multipart part's Content-Length
 * when the headers directly precede the JPEG (robust against embedded EXIF thumbnails, which
 * contain their own FFD9), otherwise cuts on SOI (FFD8) … EOI (FFD9).
 * Not thread-safe — use from the network thread only.
 */
final class MjpegParser {

    interface FrameSink {
        /** The array is owned by the receiver. */
        void onFrame(byte[] jpeg);
    }

    static final int MAX_FRAME = 4 * 1024 * 1024;
    private static final int HEADER_WINDOW = 256;

    private byte[] buf = new byte[256 * 1024];
    private int len = 0;
    private int start = -1;
    private int scanFrom = 0;

    void reset() {
        len = 0;
        start = -1;
        scanFrom = 0;
    }

    void feed(byte[] chunk, int n, FrameSink sink) {
        if (n <= 0) return;
        ensureCapacity(len + n);
        System.arraycopy(chunk, 0, buf, len, n);
        len += n;

        int consumed = 0;
        while (true) {
            if (start < 0) {
                start = find(0xD8, Math.max(consumed, scanFrom));
                if (start < 0) {
                    // Keep a tail: FF D8 may straddle chunks, and the part headers (Content-Length)
                    // that precede the next JPEG must survive until it arrives.
                    consumed = Math.max(consumed, len - HEADER_WINDOW);
                    scanFrom = Math.max(consumed, len - 1);
                    break;
                }
            }

            int declared = partLength(start);
            if (declared > 0) {
                if (len < start + declared) break;
                sink.onFrame(Arrays.copyOfRange(buf, start, start + declared));
                consumed = start + declared;
                start = -1;
                scanFrom = consumed;
                continue;
            }

            int end = find(0xD9, Math.max(start + 2, scanFrom));
            if (end < 0) {
                scanFrom = Math.max(start + 2, len - 1);
                if (len - start > MAX_FRAME) { // garbage — resync
                    consumed = len;
                    start = -1;
                    scanFrom = len;
                }
                break;
            }
            sink.onFrame(Arrays.copyOfRange(buf, start, end + 2));
            consumed = end + 2;
            start = -1;
            scanFrom = consumed;
        }
        compact(consumed);
    }

    private void compact(int consumed) {
        if (consumed <= 0) return;
        int remain = len - consumed;
        if (remain > 0) System.arraycopy(buf, consumed, buf, 0, remain);
        len = Math.max(remain, 0);
        if (start >= 0) start -= consumed;
        scanFrom = Math.max(0, scanFrom - consumed);
    }

    private int find(int second, int from) {
        for (int i = Math.max(0, from); i < len - 1; i++) {
            if ((buf[i] & 0xFF) == 0xFF && (buf[i + 1] & 0xFF) == second) return i;
        }
        return -1;
    }

    /** Content-Length from part headers that end ("\r\n\r\n") exactly where the JPEG starts. */
    private int partLength(int jpegStart) {
        if (jpegStart < 16) return 0;
        int from = Math.max(0, jpegStart - HEADER_WINDOW);
        String head = new String(buf, from, jpegStart - from, StandardCharsets.ISO_8859_1);
        int blank = head.lastIndexOf("\r\n\r\n");
        if (blank < 0 || head.length() - blank != 4) return 0;
        int boundary = Math.max(0, head.lastIndexOf("--"));
        String lower = head.substring(boundary).toLowerCase(Locale.ROOT);
        int k = lower.indexOf("content-length:");
        if (k < 0) return 0;
        int i = k + "content-length:".length();
        while (i < lower.length() && lower.charAt(i) == ' ') i++;
        int j = i;
        while (j < lower.length() && Character.isDigit(lower.charAt(j))) j++;
        if (j == i) return 0;
        try {
            int v = Integer.parseInt(lower.substring(i, j));
            return v > 0 && v < MAX_FRAME ? v : 0;
        } catch (NumberFormatException e) {
            return 0;
        }
    }

    private void ensureCapacity(int need) {
        if (need <= buf.length) return;
        int cap = buf.length;
        while (cap < need) cap *= 2;
        buf = Arrays.copyOf(buf, Math.min(cap, MAX_FRAME * 2));
        if (buf.length < need) { // pathological input: drop and resync
            reset();
            buf = new byte[256 * 1024];
        }
    }
}
